### PR TITLE
[glean] 1533746: Make events survive application restarts.

### PR DESCRIPTION
**Pull request by mdboom**
This is a second attempt at #2328 (which was reverted using #2524 since it was the cause of intermittent failures).

The first commit here is the original content of #2328 as merged.  The second commit are some fixes to address the intermittent failures.  This should hopefully make it easier to review the fixes independently from the prior work that was already approved.

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

**Comment by mdboom** [comment]
One successful run.

File: components/service/glean/src/test/java/mozilla/components/service/glean/storages/EventsStorageEngineTest.kt [hunk 1]
```diff
@@ -324,4 +339,122 @@ class EventsStorageEngineTest {
...
+        event.record(mapOf(ExtraKeys.Key1 to "bar"))
+        EventsStorageEngine.testWaitForWrites()
+        // Add a couple of truncated events to disk. One is still valid JSON, the other isn't.
+        // These event should be skipped, all others intact.
+        EventsStorageEngine.writeEventToDisk("store1", "[500]")
+        EventsStorageEngine.testWaitForWrites()
+        EventsStorageEngine.writeEventToDisk("store1", "[500, \"foo")
+        EventsStorageEngine.testWaitForWrites()
+        event.record(mapOf(ExtraKeys.Key1 to "baz"))
+        EventsStorageEngine.testWaitForWrites()
```
Comment (mdboom) [inline components/service/glean/src/test/java/mozilla/components/service/glean/storages/EventsStorageEngineTest.kt]: These additional `testWaitForWrites` is so these lines are written to the events file in order.  The failure in this test was due to checking the events in a different order in which they ended up in the file.

On this test, I'm pretty confident that these changes should help.

File: components/service/glean/src/test/java/mozilla/components/service/glean/TestUtil.kt [hunk 2]
```diff
@@ -197,9 +194,7 @@ internal fun triggerWorkManager() {
         isWorkScheduled(PingUploadWorker.PING_WORKER_TAG))
 
     // Since WorkManager does not properly run in tests, simulate the work being done
-    GlobalScope.launch(Dispatchers.IO) {
-        PingUploadWorker.uploadPings()
-    }
+    PingUploadWorker.uploadPings()
```
Comment (mdboom) [inline components/service/glean/src/test/java/mozilla/components/service/glean/TestUtil.kt]: This just seems like good low-hanging fruit, but I'm not sure if there's a clear relationship to the failing tests.

**Comment by mdboom** [comment]
Second run, failed in `service-sync`:

```
Execution failed for task ':support-base:transformClassesWithJacocoForDebug'.
> Could not resolve all files for configuration ':support-base:androidJacocoAnt'.
   > Could not download asm-debug-all.jar (org.ow2.asm:asm-debug-all:5.2)
      > Could not get resource 'https://jcenter.bintray.com/org/ow2/asm/asm-debug-all/5.2/asm-debug-all-5.2.jar'.
         > Could not HEAD 'https://d29vzk4ow07wi7.cloudfront.net/254b82bec9da4f8efbc8b1f93ab2b87f7465227a82b36cf3d05d9e77a0e8dd2e?response-content-disposition=attachment%3Bfilename%3D%22asm-debug-all-5.2.jar%22&Policy=eyJTdGF0ZW1lbnQiOiBbeyJSZXNvdXJjZSI6Imh0dHAqOi8vZDI5dnprNG93MDd3aTcuY2xvdWRmcm9udC5uZXQvMjU0YjgyYmVjOWRhNGY4ZWZiYzhiMWY5M2FiMmI4N2Y3NDY1MjI3YTgyYjM2Y2YzZDA1ZDllNzdhMGU4ZGQyZT9yZXNwb25zZS1jb250ZW50LWRpc3Bvc2l0aW9uPWF0dGFjaG1lbnQlM0JmaWxlbmFtZSUzRCUyMmFzbS1kZWJ1Zy1hbGwtNS4yLmphciUyMiIsIkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTU1MzYxODIxNn0sIklwQWRkcmVzcyI6eyJBV1M6U291cmNlSXAiOiIwLjAuMC4wLzAifX19XX0_&Signature=EHcsy1aeoBAZwu0n3u~B3BJsMe2oJxpRbREuZjXco1DKTKDesZ5PbqyK9jz-atxOcPDCB8h7sjyprnS0pqwiUbQATNOr-h7AxBkkt9mrkfQoYP6wW0RA2qW-DgaC5vvmNLIZy1yl6k4v80Ww9u~bOsO01ABS4fCckAJndAa-I71ZU5SzyVkpiCPBQm7We1o8dUbWQlUvjg6FtxTXKp9afVyup0j3sKb5M4mLvLhefCJtg0hxtuqvRmJvgJgVJbS6C9zRvpWkn9p82AJXDMfUUhMldpSM4zApAqImGYDw4hejEcaCl9RnxZqHLph1fu3uVxHGISFkGO5vc0N9rUeqbA__&Key-Pair-Id=APKAIFKFWOMXM2UMTSFA'.
            > Remote host closed connection during handshake
```

**Comment by mdboom** [comment]
Another interesting data point: `EventsStorageEngineTest` `getSnapshot() correctly clears the stores` failed in another PR that doesn't include the changes in this PR, so it's possibly due to something unrelated to these changes.

https://taskcluster-artifacts.net/b5D0VJthSYGGO_mZ3Okzvg/0/public/logs/live_backing.log

**Comment by mdboom** [comment]
Third run passing.

**Comment by mdboom** [comment]
Fourth run passing.

**Comment by Dexterp37** [comment]
Let's make sure this passes at least 5 time before merging..

**Comment by mdboom** [comment]
Unrelated failure in `samples-browser`

**Comment by mdboom** [comment]
A bunch of unrelated failures in other packages.

**Comment by mdboom** [comment]
Five successful runs (with no failing in glean -- only in other tasks)

**Comment by travis79** [comment]
Go for it

**Pull request merged**